### DISCUSSION TITLE
Module localization

### DISCUSSION THREAD

**Discussion by JanOlsmar** [oqtane/oqtane.framework#2293] [Q&A] [answered]
I can not get localization to work for my modules. That is, text in resx files is displayed for other than the original resx file.
If I create a module with Modul Creator. Adding resx files for my language and other languages I can not get these to appear.
I have other languages installed. 
So it must be something I missed to make this work.

Please help me in right direction.

**Comment by leigh-pointer** [comment, marked as answer]
How are you building the lang resource dll’s?
________________________________
From: JanOlsmar ***@***.***>
Sent: Monday, July 18, 2022 11:38:55 AM
To: oqtane/oqtane.framework ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [oqtane/oqtane.framework] Module localization (Discussion #2293)


I can not get localization to work for my modules. That is, text in resx files is displayed for other than the original resx file.
If I create a module with Modul Creator. Adding resx files for my language and other languages I can not get these to appear.
I have other languages installed.
So it must be something I missed to make this work.

Please help me in right direction.

**Reply by JanOlsmar** [reply]
The frameworks dll I use the cmd as shown in that project.
I do not build any dlls for the module resource file.

**Reply by leigh-pointer** [reply]
@JanOlsmar do you have this line appended to you debug.cmd file

where xx-XX is your language codes.

`XCOPY "..\Client\bin\Debug\net6.0\xx-XX\YouCompany.YourModule.Client.Oqtane.resources.dll" "..\..\oqtane.framework\Oqtane.Server\bin\Debug\net6.0\xx-XX\" /Y`

**Reply by JanOlsmar** [reply]
No, the cmd file is the cmd file from the module creator.

**Reply by leigh-pointer** [reply]
You will need copy the resources.dll from the debug folder to your server bin path, making sure that the target is the correct language code.

**Reply by JanOlsmar** [reply]
I will test that.  Thanks so far :)

**Reply by JanOlsmar** [reply]
That worked much better now. Thanks.  Some Oqtane controls (Add Edit etc) did not work but I will look in to it a little more.
I guess I have to add something in nuspec to get the release version to work too.

**Reply by leigh-pointer** [reply]
@JanOlsmar I create a separate nuspec for each language pack and basically it is 

YourCompany.YourModule.xx-XX.nuspec

Contents

```
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2010/07/nuspec.xsd">
  <metadata>
    <id>YourCompany.YourModule.nl-NL</id>
    <version>1.0.0</version>
    <authors>YourCompany</authors>
    <owners>YourCompany</owners>
    <title>YourCompany ModuleName</title>
    <description>Module Name</description>
    <copyright>YourCompany</copyright>
    <requireLicenseAcceptance>false</requireLicenseAcceptance>
    <license type="expression">MIT</license>
    <projectUrl>https://Your Project URL</projectUrl>
    <icon>Icon.png</icon>
    <tags>oqtane translations</tags>
    <releaseNotes></releaseNotes>
    <summary></summary>
    <dependencies>
      <dependency id="Oqtane.Framework" version="3.1.3" />
    </dependencies>
    <language>xx-XX</language>
  </metadata>
  <files>
    <file src="..\Client\bin\Release\net6.0\xx-XX\YourCompany.YourModule.Client.Oqtane.resources.dll" target="lib\net6.0\xx-XX" />
    <file src="Icon.png" target="" />
  </files>
</package>
```

**Reply by JanOlsmar** [reply]
Thanks for the help today. Made me take another step into Oqtane's world.

**Reply by sbwalker** [reply]
@leigh-pointer does it make sense to update the External Template to include all of the changes suggested above to make localization development simpler in modules?

**Reply by leigh-pointer** [reply]
@sbwalker yes that would be a good idea.  it also might be worth generating at module creation time the language resx if an localization option was selected in the Creation Module?

**Comment by hishamco** [comment]
@JanOlsmar please check [Oqtane Blogs](https://github.com/oqtane/oqtane.blogs) as reference

**Reply by JanOlsmar** [reply]
What to check there? There are no translations.

**Reply by hishamco** [reply]
There will be `Localizer` and `ResourceKey` have been used.

> What to check there? There are no translations.

Please check https://github.com/oqtane/oqtane.blogs/tree/master/Client/Resources

At the end you can copy and create once target your culture

**Reply by JanOlsmar** [reply]
Thats not my issue.  But I have a question regarding those filenames. There the name is Oqtane.Blogs.Edit.resx  direct below the resource folder. In Module creator a folder "Oqtane.Blogs" had been created and there the Edit, Index and Settings.resx. 
Is both methods valid?
(by the way Settings page use the Edit.resx)

**Reply by hishamco** [reply]
I can say Yes or No, because sometimes the namespace the assembly names make the localization broken. do you have a minimal repo that reproduce the issue?

**Reply by JanOlsmar** [reply]
[Olsmar.LocTest.zip](https://github.com/oqtane/oqtane.framework/files/9132947/Olsmar.LocTest.zip)

**Reply by hishamco** [reply]
Seems you already mark the best answer, so I think your issue has been fixed
